### PR TITLE
feat: add VICReg Implementation

### LINKED ISSUE #8: (new-method) add VICReg implementation
Add Implementation of [VICReg: Variance-Invariance-Covariance Regularization for Self-Supervised Learning)](https://arxiv.org/abs/2105.04906).

We mostly depend on the official code released on [facebookresearch/vicreg](https://github.com/facebookresearch/vicreg). 

<details>
<summary><b>Citation</b></summary>

```
@inproceedings{bardes2022vicreg,
  author  = {Adrien Bardes and Jean Ponce and Yann LeCun},
  title   = {VICReg: Variance-Invariance-Covariance Regularization For Self-Supervised Learning},
  booktitle = {ICLR},
  year    = {2022},
}
```

</details>

**Notebooks:-**

- [x] Add PyTorch Implementation (ported from the original implementation) ([344ba0b](https://github.com/SauravMaheshkar/Self-Supervised-Learning/pull/9/commits/344ba0b1da24fd9e39b76f4b119eb01a6b21eb96))
- [x] Add Tensorflow Implementation ([eed7d38](https://github.com/SauravMaheshkar/Self-Supervised-Learning/pull/9/commits/eed7d385ae93495b28e0118618541b78796dbad8))
- [ ] Add JAX Implementation (`Flax`)

**Modules:-**

- [x] PyTorch
    - [x] augmentations: add `Gaussian Blur` and `Solarization` implementations with related tests. ([8d9d770](https://github.com/SauravMaheshkar/Self-Supervised-Learning/pull/9/commits/8d9d770cb851f8725efbc5a38f6c6a6036763a54))
    - [x] optimizers: add `LARS` implementation with related tests. (d0dd63dcf608617f3cbd4d76ec342f1832f8b319)
    - [x] model: add `VICReg` implementation with related tests. (dca4ce233d9bbac3768696ae90566d8c9d00eae5)
- [x] Tensorflow
    - [x] augmentations: add `shuffle__zipped_outputs` with arbitrary # of inputs, `scale_image`, `gaussian_blur`,  `color_jitter`, `solarize`, `color_drop` and `random_apply` with related tests. (8e5ae66ce2c1ed36f61881a4918451db8853066c)
    - [x] model: add `VICReg` implementation with related tests. (daee4368ab7d10655f056cc720c45ed0f1ac2107)

### PR DESCRIPTION
**Pull request by SauravMaheshkar**
Fixes #8

### CODE REVIEW

File: notebooks/VICReg/Tensorflow_VICReg.ipynb [hunk 1]
```diff
@@ -0,0 +1,774 @@
+{
```
Comment (soumik12345) [inline notebooks/VICReg/Tensorflow_VICReg.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #25.    strategy = tf.distribute.MirroredStrategy()
</pre><p>Also add an option for TPU Strategy. The pipeline seems to be completely TPU compatible.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/SauravMaheshkar/Self-Supervised-Learning/pull/9/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='1'/>

**Review by soumik12345 — APPROVED** [review]
Hi @SauravMaheshkar I confined my review to the JAX and TensorFlow versions only. The code for the TF version looks superb. As for the JAX version, is it WIP?

**Comment by SauravMaheshkar** [comment]
Yes atm the JAX version is a WIP. I'll re-ping once the JAX notebook is ready for review. Thanks for the review on the Tensorflow implementation.

File: notebooks/VICReg/Tensorflow_VICReg.ipynb (hunk 1 above)
Comment (SauravMaheshkar) [reply notebooks/VICReg/Tensorflow_VICReg.ipynb:1]: I intend to get to TPU's in a while. First I want to benchmark PyTorch vs Tensorflow vs JAX on GPUs (I know JAX + GPU isn't that great, but I want to benchmark regardless). TPU Support would be added in the future.

**Comment by SauravMaheshkar** [comment]
<p>Makes sense, I'll move it into utilities and add better docstrings.</p>

---

 <a href='https://app.reviewnb.com/SauravMaheshkar/Self-Supervised-Learning/pull/9/discussion/'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Comment by SauravMaheshkar** [comment]
<p>Resolved in <a href="https://github.com/SauravMaheshkar/Self-Supervised-Learning/pull/9/commits/601f592f1c430016da120d40a8c7bfaf3c70224c" target="_blank">601f592</a></p>

---

 <a href='https://app.reviewnb.com/SauravMaheshkar/Self-Supervised-Learning/pull/9/discussion/'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Comment by SauravMaheshkar** [comment]
<p>Resolved in <a href="https://github.com/SauravMaheshkar/Self-Supervised-Learning/pull/9/commits/3635f4ee22c4e9570a1c7d028372ba0a4e1ebf85" target="_blank">3635f4e</a></p>

---

 <a href='https://app.reviewnb.com/SauravMaheshkar/Self-Supervised-Learning/pull/9/discussion/'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Comment by SauravMaheshkar** [comment]
<p>Resolved in <a href="https://github.com/SauravMaheshkar/Self-Supervised-Learning/pull/9/commits/17565f3744f36dac72f0f5e252dcbc65c6d89c98" target="_blank">17565f3</a></p>

---

 <a href='https://app.reviewnb.com/SauravMaheshkar/Self-Supervised-Learning/pull/9/discussion/'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Comment by SauravMaheshkar** [comment]
<p>Resolved in <a href="https://github.com/SauravMaheshkar/Self-Supervised-Learning/pull/9/commits/17565f3744f36dac72f0f5e252dcbc65c6d89c98" target="_blank">17565f3</a></p>

---

 <a href='https://app.reviewnb.com/SauravMaheshkar/Self-Supervised-Learning/pull/9/discussion/'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Comment by SauravMaheshkar** [comment]
<p>Yes, apparently it's not recommended. [Source](https://discuss.pytorch.org/t/what-is-torch-cuda-empty-cache-do-and-where-should-i-add-it/40975/2?u=sauravmaheshkar)</p>

---

 <a href='https://app.reviewnb.com/SauravMaheshkar/Self-Supervised-Learning/pull/9/discussion/'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Comment by SauravMaheshkar** [comment]
<p>Resolved in <a href="https://github.com/SauravMaheshkar/Self-Supervised-Learning/pull/9/commits/ad8b06ca112cfca63d207b0f1dbd925cda1322c1" target="_blank">ad8b06c</a></p>

---

 <a href='https://app.reviewnb.com/SauravMaheshkar/Self-Supervised-Learning/pull/9/discussion/'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Comment by SauravMaheshkar** [comment]
<p>Great suggestion. I eventually plan to expand this project to include scripts and modules as well and then introduce testing. Atm I want to create minimal notebook implementations in various frameworks. I could've included tests within the notebook itself but that would've increased clutter within the notebook making it more extensive but longer at the same time. For now I'll stick to plain notebooks but when I eventually add modules I'll surely include tests as well. (I've planned ahead and added a <code>tests/</code> dir and some pytest configuration beforehand )</p>

---

 <a href='https://app.reviewnb.com/SauravMaheshkar/Self-Supervised-Learning/pull/9/discussion/'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>
